### PR TITLE
Fix `PruneArrays[LeftShift]` when struct is `Hole`.

### PR DESCRIPTION
**Pull request by sellout**
This also refactors PruneArrays to use State and also to give more descriptive names.

### CODE REVIEW

**Comment by sellout** [comment]
NB: This doesn’t yet apply PruneArrays to branches, but prepares it.

File: connector/src/main/scala/quasar/qscript/PruneArrays.scala [hunk 1]
```diff
@@ -26,86 +25,84 @@ import matryoshka._
...
 import simulacrum.typeclass
 
 object PATypes {
-  type Acc = Set[BigInt]
-  type StateAcc = Option[Acc]
-
-  implicit final class StateAccOps(val self: StateAcc) extends AnyVal {
-    def |++|(other: StateAcc): StateAcc =
-      Semigroup.liftSemigroup[Option, Acc].append(self, other)
+  type SeenIndices = Set[BigInt]
```
Comment (alissapajer) [inline connector/src/main/scala/quasar/qscript/PruneArrays.scala]: +1 to the name change.

Comment (alissapajer) [inline connector/src/main/scala/quasar/qscript/PruneArrays.scala]: And to all the other names changes too. They are better.

File: connector/src/main/scala/quasar/qscript/PruneArrays.scala [hunk 2]
```diff
@@ -54,31 +54,37 @@ class PAHelpers[T[_[_]]: BirecursiveT: EqualT] extends TTypes[T] {
...
-          case (-\/(SrcHole), _)                       => None  // non-static index
-          case (_, _)                                  => (acc1 ++ acc2).some
+  def findIndicesInFunc[A](func: Free[MapFunc, A]): ScalaMap[A, KnownIndices] =
+    func.project.run.fold(k => ScalaMap(k -> none), κ(findIndicesInStruct(func)))
+
+  def findIndicesInStruct[A](func: Free[MapFunc, A]): ScalaMap[A, KnownIndices] = {
+    def accumulateIndices: GAlgebra[(Free[MapFunc, A], ?), MapFunc, ScalaMap[A, KnownIndices]] = {
+      case ProjectIndex((src, acc1), (value, acc2)) =>
+        val newMap = mergeMaps(acc1, acc2)
+          (src.project.run, value.project.run) match {
```
Comment (alissapajer) [inline connector/src/main/scala/quasar/qscript/PruneArrays.scala]: Minor formatting - this section is too far indented.

**Comment by alissapajer** [comment]
@sellout LGTM, pending build issues.

**Comment by alissapajer** [comment]
@sellout not sure why the compile errors aren't showing up in the travis logs, but this branch doesn't compile for me after `sbt clean`. Compile error in `PruneArrays.scala`: `not found: value extractFromMap`.

**Comment by sellout** [comment]
Whoops – how did that happen …

**Pull request merged**